# Workflow Overview
Bactopia is an extensive workflow integrating numerous steps in bacterial genome analysis. Through out the workflow there are steps that are *always enabled* and *dataset enabled*. Each of the steps depicted in the image below will be described in this section.
![Bactopia Workflow](data/bactopia-workflow.png)

## Always Enabled Steps
The *Always Enabled Steps* (as you might imagine!) are always executed by Bactopia. These steps do not depend of external datasets and thus are always enabled.

### Gather FASTQs

### Validate FASTQs

### Original Summary

### Genome Size

### Quality Control


### QC Summary

### Count 31-mers

### De novo Assembly

### Genome Annotation

### Antimicrobial Resistance

## Dataset Enabled Steps
### Public Dataset Enabled


#### Call Variants (Auto)

#### Local Assembly

#### Minmer Query

#### Sequence Type


### User Dataset Enabled
#### BLAST Alignment

#### Call Variants (User)

#### Insertion Sites

#### Reference Mapping
